pycalphad/core/calculate.py

```diff
@@ -18,6 +19,54 @@
 from pycalphad.core.utils import endmember_matrix, extract_parameters, \
     get_pure_elements, filter_phases, instantiate_models, point_sample, \
     unpack_components, unpack_condition, unpack_kwarg
+from pycalphad.core.constants import MIN_SITE_FRACTION
+
+
+def hr_point_sample(constraint_jac, constraint_rhs, initial_point, num_points):
+    "Hit-and-run sampling of linearly-constrained site fraction spaces"
+    q, r = np.linalg.qr(constraint_jac.T, mode='complete')
+    q1 = q[:, :constraint_jac.shape[0]]
+    q2 = q[:, constraint_jac.shape[0]:]
+    r1 = r[:constraint_jac.shape[0], :]
+    if initial_point is not None:
+        z_bar = initial_point
+    else:
+        # minimum norm solution to underdetermined system of equations
+        # may not be feasible if it fails the non-negativity constraint
+        z_bar = np.linalg.lstsq(constraint_jac, constraint_rhs, rcond=None)[0]
+    solution_norm = np.linalg.norm(constraint_jac.dot(z_bar) - constraint_rhs)
+    if (solution_norm > 1e-4) or np.any(z_bar < 0):
+        # initial point does not satisfy constraints; give up
+        return np.empty((0, z_bar.shape[0]))
+    # Hit-and-Run sampling
+    new_feasible_z = np.zeros((num_points, constraint_jac.shape[1]))
+    current_z = np.array(z_bar)
+    min_z = MIN_SITE_FRACTION
+    rng = np.random.RandomState(1769)
```

Comment:
Only available since numpy 1.17: https://albertcthomas.github.io/good-practices-random-number-generators/
Looks like the main difference is they changed the PRNG, and NEP 19 says the `RandomState` API will continue working indefinitely: https://numpy.org/neps/nep-0019-rng-policy.html

Do you think it's worth bumping the numpy minimum version?

Reply:
I forgot how old we had runtime support for. If this isn't being deprecated anytime soon, then I don't think we need to bump the dependency. But I will note that NumPy's [NEP-29](https://numpy.org/neps/nep-0029-deprecation_policy.html) suggests that community support NumPy versions from the last 24 months of minor releases, so as of today NumPy 1.19.